pkg/providers/instancetype/suite_test.go

```diff
@@ -1716,6 +1718,159 @@ var _ = Describe("InstanceTypes", func() {
+			// kubelet.maxPods
+			nodePool.Spec.Template.Spec.Kubelet = &corev1beta1.KubeletConfiguration{
+				KubeReserved:   v1.ResourceList{v1.ResourceCPU: resource.MustParse("1")},
+				SystemReserved: v1.ResourceList{v1.ResourceCPU: resource.MustParse("1")},
+				EvictionHard:   map[string]string{"memory.available": "5%"},
+				EvictionSoft:   map[string]string{"nodefs.available": "10%"},
+				MaxPods:        aws.Int32(10),
+			}
+			kubeletChanges := []*corev1beta1.KubeletConfiguration{
+				{KubeReserved: v1.ResourceList{v1.ResourceCPU: resource.MustParse("20")}},
+				{SystemReserved: v1.ResourceList{v1.ResourceMemory: resource.MustParse("10Gi")}},
+				{EvictionHard: map[string]string{"memory.available": "52%"}},
+				{EvictionSoft: map[string]string{"nodefs.available": "132%"}},
+				{MaxPods: aws.Int32(20)},
+			}
+			var instanceTypeResult [][]*corecloudprovider.InstanceType
+			ExpectApplied(ctx, env.Client, nodeClass)
+			// Adding the general set of to the instancetype into the cache
+			fullInstanceTypeList, err := cloudProvider.GetInstanceTypes(ctx, nodePool)
+			Expect(err).To(BeNil())
+			sort.Slice(fullInstanceTypeList, func(x int, y int) bool {
+				return fullInstanceTypeList[x].Name < fullInstanceTypeList[y].Name
+			})
+
+			sorted := nodePool.DeepCopy()
+			for _, change := range kubeletChanges {
+				nodePool = sorted.DeepCopy()
+				Expect(mergo.Merge(nodePool.Spec.Template.Spec.Kubelet, change, mergo.WithOverride, mergo.WithSliceDeepCopy)).To(BeNil())
+				// Calling the provider and storing the instance type list to the instancetype provider cache
+				_, err := awsEnv.InstanceTypesProvider.List(ctx, nodePool.Spec.Template.Spec.Kubelet, nodeClass)
+				Expect(err).To(BeNil())
+				// We are making sure to pull from the cache
+				instanetypes, err := awsEnv.InstanceTypesProvider.List(ctx, nodePool.Spec.Template.Spec.Kubelet, nodeClass)
+				Expect(err).To(BeNil())
+				sort.Slice(instanetypes, func(x int, y int) bool {
+					return instanetypes[x].Name < instanetypes[y].Name
+				})
+				instanceTypeResult = append(instanceTypeResult, instanetypes)
+			}
+
+			// We expect that the updated kubelet configuration will result in a subset of the full set instance types thats offered
+			// The goal is to make sure that we don't have different configuration collapsing to the same cache key and have a fewer set of instancesTypes then expected
+			for _, it := range instanceTypeResult {
+				Expect(reflect.DeepEqual(fullInstanceTypeList, it)).To(BeFalse())
+			}
+			// Based on the nodeclass configuration, we expect to have 5 unique set of instance types
+			for x := range instanceTypeResult {
+				for y := range instanceTypeResult {
+					if x == y {
+						continue
+					}
+					key := reflect.DeepEqual(instanceTypeResult[x], instanceTypeResult[y])
+					if key {
+						fmt.Printf("failed x: %d, y: %d", x, y)
+					}
+					Expect(reflect.DeepEqual(instanceTypeResult[x], instanceTypeResult[y])).To(BeFalse())
+				}
+			}
```

Comment:
is this something that can be done with `lo.Uniq()`? Or does that not do DeepEquals?

Reply:
`lo.Uniq()` needs a type that's comparable and []InstanceType is not a really comparable type so we need to do a `DeepEquals`